Tdd tests

bertil291utn opened the pull request:
### Testing TDD with RSpec in enumerable methods
We tested the following methods
- My_each
- My_each_with_index
- My_select
- My_all?
- My_any?
- My_none?
- My_count
- My_map
- My_inject

It was tested with different kind of variables (procs, range, arrays) and different situations


We really appreciate your feedback, thank you in beforehand 💻 🚀

bertil291utn committed:
Init rspec tdd test files

bertil291utn committed:
Add rufo file to gitignore

bertil291utn committed:
My each method
- Test if the original array is equal to result array  in
my each method

bertil291utn committed:
My each method
- Add two tests to my_each method array (string and numbers)

bertil291utn committed:
My each with index method
- Add block and block no given cases
- Add no given method to my_each method

bertil291utn committed:
Add  `describe` block every method
- Add an mix array whit nil and false values

bertil291utn committed:
Add my all, my_inject, my none, my_any, my_count
and my_map methods in
rsepec file

bertil291utn committed:
Fix condition comparation
- Add a mock result instead a fixed result

bertil291utn committed:
Disable rubocop in spec_ helper fie

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+  let(:my_proc_map) { proc { |elem| elem * 10 } }
+
+  describe '#my_each' do
+    it 'is returning same array as original number one' do
+      result = []
+      array_numbers.my_each { |elem| result.push(elem) }
+      expect(array_numbers).to eql(result)
+    end
+
+    it 'is returning same array as original string one' do
+      result = []
+      array_string.my_each { |elem| result.push(elem) }
+      expect(array_string).to eql(result)
+    end
```
[Optional]
As it is testing the same case, it returns the same array or not(Does not depend on the type of value it contains). You can combine it withing one test case. 
```ruby
it 'is returning same array as original one' do
      result = []
      array_numbers.my_each { |elem| result.push(elem) }
      expect(array_numbers).to eql(result)
      result = []
      array_string.my_each { |elem| result.push(elem) }
      expect(array_string).to eql(result)
end
```

bertil291utn replied:
Thanks for the feedback, this bug was solved you said as optional  but anyways it's done

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+      expect(result).to be_an Enumerator
+    end
+
+    it 'returns the same array as original mix array' do
+      result = []
+      array_mix.my_each { |elem| result.push(elem) }
+      expect(array_mix).to eql(result)
+    end
```
[Optional]
Well, you can test it with the original one. As you have covered the function with the array previously. Something like,
```ruby
it 'returns the same result as the original each does' do
      calculated_result = []
      array_mix.my_each { |elem| calculated_result.push(elem) }
      result = []
      array_mix.each { |elem| result.push(elem) }
      expect(calculated_result).to eql(result)
end
```

bertil291utn replied:
Thanks for the feedback, this bug was solved you said as optional but anyways it's done

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+      expect(my_result).to eql([10, 3, 5, 7, 5, 10])
+    end
+
+    it 'is returning same array as original string one' do
+      my_result = []
+      array_string.my_each_with_index { |elem, i| my_result.push(elem + i.to_s) }
+      expect(my_result).to eql(%w[100 21 32 43 14 55 arbol6])
+    end
```
[Optional]
You can also compare it with the original method by following the previous commented code.

bertil291utn replied:
Thanks for the feedback, this bug was solved you said as optional but anyways it's done

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+  end
+
+  describe '#my_select' do
+    it 'returns an enumerator value' do
```
When it returns an enumerator? I can not get a message from the test.(it should be like `when no block is given, returns an enumerator`).

bertil291utn replied:
Thanks for the feedback, this bug was solved

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+    end
+
+    it 'returns a new array with elements no equal to true' do
+      expect(array_mix.my_select { |elem| elem }).not_to eql([nil, 99])
```
Why it should not contain 99? I tried to execute it with the original select 
```ruby
[nil, true, 99].select {|element| element}   #Returned me [true, 99]
```

bertil291utn replied:
Yeah certainly 99 value should be returned and it was changed to 
```
 it 'returns a new array with elements no equal to true' do
      expect(array_mix.my_select { |elem| elem }).not_to eql([nil])
```
Thanks for the feedback, this bug was solved.

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+      expect(result).to be_an Enumerator
+    end
+
+    it 'returns a new array with elements equal to 1' do
+      expect(array_numbers.my_select { |elem| elem == 1 }).to eql([1])
+    end
+
+    it 'returns a new array with elements no equal to true' do
+      expect(array_mix.my_select { |elem| elem }).not_to eql([nil, 99])
+    end
```
I just refactored your tests in the format below. Now it seems more readable to me.
```ruby
context 'when block given' do
      it 'returns a new array containing the elements that matches the block' do
        expect(array_numbers.my_select { |elem| elem == 1 }).to eql([1])
      end

      it 'returns a new array that does contain nil/false' do
        expect(array_mix.my_select { |elem| elem }).to eql(array_mix.select { |elem| elem })
      end
end
```
The output seems like.
![image](https://user-images.githubusercontent.com/38625956/79689006-f87ccd80-8273-11ea-8c44-647ae9c97580.png)

bertil291utn replied:
Thanks for the feedback, this bug was solved, adding the context block

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+  end
+
+  describe '#my_all?' do
+    it 'returns true if all elements in array are greater than ten' do
+      expect(array_numbers.my_all?(&my_proc_numbers)).to be false
+    end
+
+    it 'returns false because all elements don\'t start with letter a' do
+      expect(array_string.my_all?(&my_proc_string)).to be false
+    end
+
+    it 'returns true if there are no nil or false elements on given array' do
+      expect(array_numbers.my_all?).to be true
+    end
+
+    it 'returns false if at least there are a nil or false element on given array' do
+      expect(array_mix.my_all?).to be false
+    end
```
I can not see any test case that checks for regexp or class. Please add them.

bertil291utn replied:
Thanks for the feedback this bug was solved, adding context/it blocks containing regexp and class check ones

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+  end
+
+  describe '#my_any?' do
+    it 'returns false if at least there\'s one element in array is greater than ten' do
+      expect(array_numbers.my_any?(&my_proc_numbers)).to be false
+    end
+
+    it 'returns true because there\'s an element which start with letter a' do
+      expect(array_string.my_any?(&my_proc_string)).to be true
+    end
+
+    it 'returns true if at least there\'s an element' do
+      expect(array_numbers.my_any?).to be true
+    end
+
+    it 'returns true if there are a nil or false element on given array' do
+      expect(array_mix.my_any?).to be true
+    end
```
Same as above, not test related to regexp or class

bertil291utn replied:
Thanks for the feedback this bug was solved, adding context/it blocks containing regexp and class check ones

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+  end
+
+  describe '#my_none?' do
+    it 'returns true beaca there\'s any element in array is greater than ten' do
+      expect(array_numbers.my_none?(&my_proc_numbers)).to be true
+    end
+
+    it 'returns false because there\'s just an element which start with letter a' do
+      expect(array_string.my_none?(&my_proc_string)).to be false
+    end
+
+    it 'returns false due to at least there\'s an element' do
+      expect(array_numbers.my_none?).to be false
+    end
+
+    it 'returns true due to at least there are a nil or false element ' do
+      expect([nil, false].my_none?).to be true
+    end
+
+    it 'returns true due to is an empty array ' do
+      expect([].my_none?).to be true
+    end
```
Nothing related to regexp or class.

bertil291utn replied:
Thanks for the feedback this bug was solved, adding context/it blocks containing regexp and class check ones

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+  end
+
+  describe '#my_count' do
+    it 'returns size array (6) due to there\'s no given block' do
```
It should be generic, for example->
```ruby
it 'returns the number of elements in the array' do
end
```

bertil291utn replied:
Thanks for the feedback this bug was solved, changing the block message

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+      expect(array_numbers.my_map).to be_an Enumerator
+    end
+
+    it 'returns same array as original with given block' do
+      expect(array_numbers.my_map { |elem| elem * 10 }).to eq([100, 20, 30, 40, 10, 50])
+    end
+
+    it 'returns same array as original with given proc' do
+      expect(array_numbers.my_map(&my_proc_map)).to eq([100, 20, 30, 40, 10, 50])
+    end
```
You can enclose them inside a group using the context shown above.

bertil291utn replied:
Thanks for the feedback this bug was solved, adding context block and grouping two `it` blocks

shshamim63 commented on spec/enumerable_spec.rb:
```diff
@@ -0,0 +1,184 @@
+      expect(array_numbers.my_inject(2, :+)).to eq(27)
+    end
+
+    it 'returns all additioned elements in the range plus two' do
+      expect((1..20).my_inject(2, :+)).to eq(212)
+    end
```
[Optional]
You can compare it using the original inject method.
```ruby
it 'returns the sum of all the elements in the range along with the given initial value' do
      expect((1..20).my_inject(2, :+)).to eq((1..20).inject(2, :+))
end
```

bertil291utn replied:
Thanks for the feedback also this bug was solved, adding original method as compare condition

shshamim63 reviewed (changes requested):
## Status :traffic_light: Changes Required
___
Hi @bertil291utn :wave: ,
Well done :clap: , you have added almost all the cases, but few are left behind and those are blocking me to approve it now. So let's follow the comments to complete it.
___
### :spades: Mandatory Changes :spades: 
___
1. I can not see anything related to testing in the readme. How to execute the test is not in there. So please add the info.
2. Some of the cases are left behind, mainly with my_all?, my_any?, my_none?. So please add them.
3. Some of your messages are not self-descriptive. I have added some comments to fix them. So please organize the context of your test.
___
### :clubs: Optional Changes :clubs: 
___
1. It is always better to test your methods with the original method. So if any of your methods contains a bug, you can identify it easily.
___
#### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

bertil291utn committed:
Fix first code review
- Add cases was left behind, mainly with my_all?, my_any?, my_none?
-  Update and add some messages weren't self-descriptive enough

bertil291utn committed:
Update README file
- Add testing section, install and run files

bertil291utn committed:
Update README file

majovanilla reviewed (approved):
Good job implementing the required changes!! :+1: 

Your project looks great and you are good to go and merge!!

### Project APPROVED

You can submit a project completion form, now.

Happy coding 🤓

--Maria Reyes
